:sparkles: MLIR - Add missing Quantum Computation gates in Round Trip Pass

BertiFlorea opened the pull request:
<!--- This file has been generated from an external template. Please do not modify it directly. -->
<!--- Changes should be contributed to https://github.com/munich-quantum-toolkit/templates. -->

## Description

Fixes #1067 <!--- Replace `(issue)` with the issue number fixed by this pull request. If this PR does not fix an issue, please remove this line. -->

This PR is WIP; therefore, it is marked as a draft. There are still some TODOs left inside `ToQuantumComputationPattern.cpp` and in `to-quantum-computation.mlir`, which I will address in the next couple of days.

Also, `FromQuantumComputationPattern.cpp` will have to be modified accordingly to support all the remaining gates before merging.

## Checklist:

<!---
This checklist serves as a reminder of a couple of things that ensure your pull request will be merged swiftly.
-->

- [ ] The pull request only contains commits that are focused and relevant to this change.
- [ ] I have added appropriate tests that cover the new/changed functionality.
- [ ] I have updated the documentation to reflect these changes.
- [ ] I have added entries to the changelog for any noteworthy additions, changes, fixes, or removals.
- [ ] I have added migration instructions to the upgrade guide (if needed).
- [ ] The changes follow the project's style guidelines and introduce no new warnings.
- [ ] The changes are fully tested and pass the CI checks.
- [ ] I have reviewed my own code changes.

burgholzer commented:
@BertiFlorea may I ask for an update here?

burgholzer commented:
> Hi @burgholzer, I am currently adjusting the methods that require processing the two target gates. Specifically, in `ToQuantumComputationPattern.cpp`, the specific methods would be `handleUnitaryOp`, `matchAndRewrite` and maybe `findQubitIndex`. In `FromQuantumComputationPattern.cpp`, there are only two minor places where adjustments are left.
> 
> I am working locally on this, but it is rather hard to debug, because you cannot be sure what caused the conversion to fail. After updating the necessary places, this should be ready for review (by the end of the week, I hope).

Thanks for the update.
In order to make it easier to debug, I'd advise writing real "unit" tests that simply test individual conversions.
That should greatly help to isolate potential errors and should give us some good coverage for the entire functionality.
Thera can never be enough tests ;o)

BertiFlorea commented:
Current status:
The first direction (ToQC) seems to generate the correct circuit when testing locally. However, I am still working on fixing the other direction (FromQC). You can see, for example, the failing CI for Linux where the circuit is printed.

Just my two cents, I don't know if you are interested, but I agree with @burgholzer that the RoundTrip pass could benefit from a small rework (as described in #1085). I noticed an issue while working on this PR, that is, that the pass could only handle single-target gates. I am trying to add the logic for multi-target gates by making changes incrementally (first make all single target gates work, then all two target, and lastly the multi target case) as discussed with @flowerthrower.

Note: The following line was useful for debugging and should be removed in the end. However, it only prints when an error happens (apparently, if the tests pass, the QC circuit is not printed at all by `llvm::outs`, therefore I had to find an alternative).
```
op.emitWarning() << circuitString << '\n';
```

burgholzer commented:
@BertiFlorea may I ask for another update here?

BertiFlorea commented:
@burgholzer, I have pushed all I had locally at the moment.

Current status: 
- [x] Single and two target gates are working as expected 🎉
- [ ] Only parameterised gates are not working.

I am working on finding a way to extract the parameters from our operations and convert them to a `double` so that they can be passed to the QuantumComputation IR. (This is the ToQC direction)

denialhaag commented:
Thank you for your work on this so far!

> I am working on finding a way to extract the parameters from our operations and convert them to a `double` so that they can be passed to the QuantumComputation IR. (This is the ToQC direction)

Some references for this:

1. The pattern for merging rotation gates has examples for accessing the parameters. See, for example, here: https://github.com/munich-quantum-toolkit/core/blob/main/mlir/lib/Dialect/MQTOpt/Transforms/MergeRotationGatesPattern.cpp#L123
2. I implemented a function for casting MLIR values to `double` in https://github.com/munich-quantum-toolkit/core/pull/1019/commits/e71748cdd8e5a0791356f0e95f4c8c6734788a28. However, we ended up not going with this, so I cannot guarantee that the function works perfectly, but maybe it's a good starting point.

denialhaag commented on mlir/test/Dialect/MQTOpt/Transforms/to-quantum-computation.mlir:
```diff
@@ -86,3 +84,125 @@ module {
+// This test checks if single target unitary gates are parsed correctly by the QuantumComputation roundtrip pass.
+module {
+    // CHECK-LABEL: func.func @testSingleTargetUnitaryGates()
+    func.func @testSingleTargetUnitaryGates() -> (!mqtopt.QubitRegister, i1) {
```
For the pattern for merging rotation gates, we went even more unit-testy and added one test per gate (see [here](https://github.com/munich-quantum-toolkit/core/blob/main/mlir/test/Dialect/MQTOpt/Transforms/merge-rotation-gates.mlir)). I think this granularity makes it a bit easier to find potential problems.

denialhaag commented on mlir/lib/Dialect/MQTOpt/Transforms/FromQuantumComputationPattern.cpp:
General comment here: Maybe it makes sense to have a brief look at the pattern for merging rotation gates (I know that I'm mentioning a lot). I ran into the same problem that you need to copy and paste a lot of code, and we created a template function that avoided that quite nicely (see [here](https://github.com/munich-quantum-toolkit/core/blob/main/mlir/lib/Dialect/MQTOpt/Transforms/MergeRotationGatesPattern.cpp#L115)). The template function definitely requires some modifications, but it should also work for what you are doing here.

denialhaag reviewed (changes requested):
I know that this is still WIP, but I already wanted to leave some feedback. It is already going in a good direction!

denialhaag commented on mlir/lib/Dialect/MQTOpt/Transforms/ToQuantumComputationPattern.cpp:
```diff
@@ -159,24 +159,98 @@ struct ToQuantumComputationPattern final : mlir::OpRewritePattern<AllocOp> {
                        std::vector<mlir::Value>& currentQubitVariables) const {
 
     // Add the operation to the QuantumComputation.
+    bool twoTarget = false;
```
All the operations have this trait, as you can see [here](https://github.com/munich-quantum-toolkit/core/blob/main/mlir/include/mlir/Dialect/Common/IR/StdOps.td.inc).

denialhaag reviewed (changes requested):
Thank you for your continued work on this! :)

Below are a couple of inline suggestions to align this PR with recent changes and best practices.

A more general comment, I think it would be good if you rebased your branch. Quite a few things have changed in the last weeks:

- https://github.com/munich-quantum-toolkit/core/pull/1099 added a translation from `QuantumComputation` to `MQTRef`. `XXminusYY` and `XXplusYY` have been renamed. Furthermore, the PR established some best practices, so it might be useful to have a brief look.
- https://github.com/munich-quantum-toolkit/core/pull/1116 and https://github.com/munich-quantum-toolkit/core/pull/1139 implemented new functions that should help reduce boilerplate code in the tests. Furthermore, https://github.com/munich-quantum-toolkit/core/pull/1139 defined custom assembly formats that should help make the tests less verbose.

BertiFlorea commented:
@denialhaag, thank you for the review and for pointing out some helpful spots!

I've solved all merge conflicts (somehow rebase didn't work as expected 🤔) and addressed all comments. Please let me know if anything else should be added or modified apart from the above! 😄 

P.S. I do not understand why the code coverage doesn't detect the test cases that use the new macro; also, most of the code highlighted is covered by test cases. Do you know maybe why that happens?

denialhaag commented:
> @denialhaag, thank you for the review and for pointing out some helpful spots!
> 
> I've solved all merge conflicts (somehow rebase didn't work as expected 🤔) and addressed all comments. Please let me know if anything else should be added or modified apart from the above! 😄

I will have another look later today!

> P.S. I do not understand why the code coverage doesn't detect the test cases that use the new macro; also, most of the code highlighted is covered by test cases. Do you know maybe why that happens?

I think I see the problem: As I asked you, you are now using `mqtopt.qubit` in your tests. However, your pattern matches the `AllocOp` (see, for example, [here](https://github.com/munich-quantum-toolkit/core/blob/da5b096321b3d0d662e2f0bfe400a0881d8be12a/mlir/lib/Dialect/MQTOpt/Transforms/ToQuantumComputationPattern.cpp#L341). As a result, the tests actually don't use your pass.

I see two solutions:

1. Update the pattern to also match `mqtopt.qubit`.
2. Don't use `mqtopt.qubit` in the tests. I'm sorry if that is the solution. I hadn't thought about this when I suggested you use `mqtopt.qubit`.

Does @burgholzer have an opinion here?

BertiFlorea commented:
I do not have anything against the second solution (I encourage it in the short term). We could keep this PR shorter and maybe tackle point 1 of Daniel's suggestion, as well as #1085, in a separate PR. 🙌🏼

denialhaag commented:
Note that the Windows CI will likely be green once #1141 is merged.

denialhaag commented on mlir/include/mlir/Dialect/Common/IR/StdOps.td.inc:
```diff
@@ -309,7 +309,7 @@ def RZXOp : UnitaryOp<"rzx", [TwoTarget, OneParameter]> {
 }
 
-def XXminusYYOp : UnitaryOp<"xxminusyy", [TwoTarget, TwoParameters]> {
+def XXminusYYOp : UnitaryOp<"xx_minus_yy", [TwoTarget, TwoParameters]> {
```
And could you also add an entry to the changelog for this?

BertiFlorea commented:
@denialhaag once again, thanks for all the advice on this one! Please let me know if something needs to be changed.

denialhaag reviewed (commented):
This looks good to me now. I'll hand over to @burgholzer for a final review. 

Edit: Maybe it would be good to update the PR description. :D
